src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,617 @@
+"""
+Contains callable operators representing functions to be called with other
+operators as input arguments.
+Contains also a decorator class for callables, which transforms them automatically in the
+specified operator function type.
+"""
+from __future__ import annotations
+
+import abc
+from functools import partial
+from typing import Callable, List, Type, Union
+
+import numpy as np
+import scipy.sparse as sps
+
+import porepy as pp
+from porepy.numerics.ad.forward_mode import Ad_array
+
+from .operators import Operation, Operator, Tree
+
+__all__ = [
+    "Function",
+    "ConstantFunction",
+    "DiagonalJacobianFunction",
+    "InterpolatedFunction",
+    "ADmethod",
+]
+
+### BASE CLASSES ------------------------------------------------------------------------------
+
+
+class AbstractFunction(Operator):
+    """Abstract class for all operator functions, i.e. functions evaluated on some other AD
+    operators.
+
+    Implements the callable-functionality and provides abstract methods for obtaining function
+    values and the Jacobian.
+    The abstraction intends to provide means for approximating operators, where values are
+    e.g. interpolated and the Jacobian is approximated using FD.
+
+    IMPLEMENTATION NOTE:
+        One can flag the operator as ``ad_compatible``. If flagged, the AD framework passes
+        AD arrays directly to the callable ``func`` and will **not** call the abstract methods
+        for values and the Jacobian during operator parsing.
+        If for some reason one wants to flag the function as AD compatible, but still have the
+        abstract methods called, this is as of now **not** supported.
+
+        For now only one child class, porepy.ad.Function, flags itself always as AD compatible.
+
+    """
+
+    def __init__(
+        self,
+        func: Callable,
+        name: str,
+        array_compatible: bool = False,
+        ad_compatible: bool = False,
+    ):
```

Comment:
@keileg I have changed the signature and added extensive documentation as discussed.
I also resolved the outdated conversation on the issue with the hierarchy. We can use this thread now, since we boiled down the problem to proper documentation of intended use.

Reply:
I think this is fair. The whole thing is still a bit opaque to me, but I believe that just reflects the inner workings of the AD framework in general, and of the functions in particular.